ext/spl/spl_directory.stub.php

```diff
@@ -6,133 +6,133 @@ class SplFileInfo implements Stringable
 
-    /** @return string|false */
-    public function getLinkTarget() {}
+    /** @tentative-return-type */
+    public function getLinkTarget(): string|false {}
 
-    /** @return string|false */
-    public function getRealPath() {}
+    /** @tentative-return-type */
+    public function getRealPath(): string|false {}
 
-    /** @return SplFileInfo */
-    public function getFileInfo(?string $class = null) {}
+    /** @tentative-return-type */
+    public function getFileInfo(?string $class = null): SplFileInfo {}
 
-    /** @return SplFileInfo|null */
-    public function getPathInfo(?string $class = null) {}
+    /** @tentative-return-type */
+    public function getPathInfo(?string $class = null): ?SplFileInfo {}
 
     /**
      * @param resource|null $context
-     * @return SplFileObject
+     * @tentative-return-type
      */
-    public function openFile(string $mode = "r", bool $useIncludePath = false, $context = null) {}
+    public function openFile(string $mode = "r", bool $useIncludePath = false, $context = null): SplFileObject {}
 
-    /** @return void */
-    public function setFileClass(string $class = SplFileObject::class) {}
+    /** @tentative-return-type */
+    public function setFileClass(string $class = SplFileObject::class): void {}
 
-    /** @return void */
-    public function setInfoClass(string $class = SplFileInfo::class) {}
+    /** @tentative-return-type */
+    public function setInfoClass(string $class = SplFileInfo::class): void {}
 
     /** @implementation-alias SplFileInfo::getPathname */
     public function __toString(): string {}
 
-    /** @return array */
-    public function __debugInfo() {}
+    /** @tentative-return-type */
+    public function __debugInfo(): array {}
 
-    /** @return void */
-    final public function _bad_state_ex() {}
+    /** @tentative-return-type */
+    final public function _bad_state_ex(): void {}
 }
 
 class DirectoryIterator extends SplFileInfo implements SeekableIterator
 {
     public function __construct(string $directory) {}
 
-    /** @return string */
-    public function getFilename() {}
+    /** @tentative-return-type */
+    public function getFilename(): string {}
 
-    /** @return string */
-    public function getExtension() {}
+    /** @tentative-return-type */
+    public function getExtension(): string {}
 
-    /** @return string */
-    public function getBasename(string $suffix = "") {}
+    /** @tentative-return-type */
+    public function getBasename(string $suffix = ""): string {}
 
-    /** @return bool */
-    public function isDot() {}
+    /** @tentative-return-type */
+    public function isDot(): bool {}
 
-    /** @return void */
-    public function rewind() {}
+    /** @tentative-return-type */
+    public function rewind(): void {}
 
-    /** @return bool */
-    public function valid() {}
+    /** @tentative-return-type */
+    public function valid(): bool {}
 
     /** @return int|false */
     public function key() {}
```

Comment:
Leave a TODO on these? I assume these are not converted because we really should be getting rid of that `false` return, right?

Reply:
I fixed this one in https://github.com/php/php-src/commit/ac7a28189abdce17816d2913d412a9d4fc3dfe0f. The `false` only happened for uninitialized objects, so we can definitely throw here.

Reply:
my problem with the `key()` method is that `DirectoryIterator` returns an `int`, while its child, `FilesystemIterator` returns a `string`. I left the parent as-is, but I can change the latter one to a real `string` return type as per your other comment.

The reason why I didn't convert the `current()` method is because DirectoryIterator::current() is annotated to return `DirectoryIterator`, while `FilesystemIterator::current()` overrides it to return `string|SplFileInfo|FilesystemIterator`.

Reply:
Distantly related: `SplFileInfo::current()` can also return `false` when there is no more lines in the file. I'm not sure if we can also change this.

Reply:
> my problem with the `key()` method is that `DirectoryIterator` returns an `int`, while its child, `FilesystemIterator` returns a `string`. I left the parent as-is, but I can change the latter one to a real `string` return type as per your other comment.
> 
> The reason why I didn't convert the `current()` method is because DirectoryIterator::current() is annotated to return `DirectoryIterator`, while `FilesystemIterator::current()` overrides it to return `string|SplFileInfo|FilesystemIterator`.

Wow. And here I thought I already knew all the terrible things SPL does. Let's leave the methods on DirectoryIterator alone then. But I think it's okay to annotate `FilesystemIterator::current()`.

Reply:
> Distantly related: `SplFileInfo::current()` can also return `false` when there is no more lines in the file. I'm not sure if we can also change this.

Unfortunately it can also return false via fgetcsv, so I don't think we can get rid of it.